packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts

```diff
@@ -85,7 +79,11 @@ function getWrappedShadowRootListener(
     let shadowRootWrappedListener = shadowRootEventListenerMap.get(listener);
     if (isUndefined(shadowRootWrappedListener)) {
         shadowRootWrappedListener = function (event: Event) {
-            if (shouldInvokeShadowRootListener(event)) {
+            const hostElement = eventCurrentTargetGetter.call(event) as HTMLElement;
+            const actualCurrentTarget = getShadowRoot(hostElement);
```

Comment:
Something looks strange here. The event listeners are cached by their function identity, however, the actual listener is always invoked with `sr` (line 87). So if the same listener is used on multiple shadow roots, the `this` value will always the first shadow root.

Example: https://jsfiddle.net/kemuapsg/1/

The listeners should be invoked with `actualCurrentTarget`.

Reply:
Nice catch, looks like `getWrappedShadowRootListener` and `getWrappedCustomElementListener` only need the `listener` argument. @Sumitjg can you add test coverage for this?

Reply:
yes, will add.